File: app/lib/models/pipeline.dart
class Pipeline {
  final String pipelineKey;
  final String repoPath;
  final int? issueNumber;
  final String? issueTitle;
  final String? taskDescription;
  final String status;
  final String? plan;
  final String? branch;
  final int? prNumber;
  final String? prUrl;
  final String? error;
  final int reviewRound;
  final String model;
  final int? templateId;
  final String? templateName;
  final double totalCostUsd;
  final DateTime createdAt;
  final DateTime updatedAt;

  Pipeline({
    required this.pipelineKey,
    required this.repoPath,
    this.issueNumber,
    this.issueTitle,
    this.taskDescription,
    required this.status,
    this.plan,
    this.branch,
    this.prNumber,
    this.prUrl,
    this.error,
    this.reviewRound = 0,
    required this.model,
    this.templateId,
    this.templateName,
    this.totalCostUsd = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Pipeline.fromJson(Map<String, dynamic> json) {
    return Pipeline(
      pipelineKey: json['pipeline_key'],
      repoPath: json['repo_path'] ?? '',
      issueNumber: json['issue_number'],
      issueTitle: json['issue_title'],
      taskDescription: json['task_description'],
      status: json['status'] ?? 'planning',
      plan: json['plan'],
      branch: json['branch'],
      prNumber: json['pr_number'],
      prUrl: json['pr_url'],
      error: json['error'],
      reviewRound: json['review_round'] ?? 0,
      model: json['model'] ?? 'sonnet',
      templateId: json['template_id'],
      templateName: json['template_name'],
      totalCostUsd: (json['total_cost_usd'] ?? 0).toDouble(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Pipeline copyWith({
    String? status,
    String? plan,
    String? branch,
    int? prNumber,
    String? prUrl,
    String? error,
    int? reviewRound,
    double? totalCostUsd,
  }) {
    return Pipeline(
      pipelineKey: pipelineKey,
      repoPath: repoPath,
      issueNumber: issueNumber,
      issueTitle: issueTitle,
      taskDescription: taskDescription,
      status: status ?? this.status,
      plan: plan ?? this.plan,
      branch: branch ?? this.branch,
      prNumber: prNumber ?? this.prNumber,
      prUrl: prUrl ?? this.prUrl,
      error: error ?? this.error,
      reviewRound: reviewRound ?? this.reviewRound,
      model: model,
      templateId: templateId,
      templateName: templateName,
      totalCostUsd: totalCostUsd ?? this.totalCostUsd,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }

  String get displayTitle =>
      issueTitle ?? taskDescription ?? 'Pipeline $pipelineKey';

  bool get isActive => const {
        'planning', 'planned', 'developing', 'developed', 'reviewing',
      }.contains(status);
}

File: backend/app/models/pipeline.py
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Pipeline(Base):
    __tablename__ = "pipelines"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    pipeline_key: Mapped[str] = mapped_column(String(100), unique=True, index=True)
    repo_path: Mapped[str] = mapped_column(String(500))
    issue_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    issue_title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    task_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(30), default="planning")
    plan: Mapped[str | None] = mapped_column(Text, nullable=True)
    branch: Mapped[str | None] = mapped_column(String(200), nullable=True)
    pr_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    pr_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    review_round: Mapped[int] = mapped_column(Integer, default=0)
    model: Mapped[str] = mapped_column(String(50), default="claude-opus-4-6")
    total_cost_usd: Mapped[float] = mapped_column(Float, default=0.0)
    claude_session_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    step_logs: Mapped[list | None] = mapped_column(JSONB, nullable=True, default=list)
    current_node_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    template_id: Mapped[int | None] = mapped_column(
        ForeignKey("pipeline_templates.id"), nullable=True
    )
    template = relationship("PipelineTemplate", lazy="joined")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
